draft-ietf-masque-h3-datagram.md

```diff
@@ -577,12 +283,22 @@ size advertised on that connection is too low), the intermediary SHOULD drop
 the HTTP Datagram instead of converting it to a DATAGRAM capsule. This
 preserves the end-to-end unreliability characteristic that methods such as
 Datagram Packetization Layer Path MTU Discovery (DPLPMTUD) depend on
-{{?RFC8899}}. An intermediary that converts QUIC DATAGRAM frames to datagram
+{{?RFC8899}}. An intermediary that converts QUIC DATAGRAM frames to DATAGRAM
 capsules allows HTTP Datagrams to be arbitrarily large without suffering any
 loss; this can misrepresent the true path properties, defeating methods such a
 DPLPMTUD.
 
 
+## The RELIABLE_DATAGRAM Capsule
+
+This document defines the RELIABLE_DATAGRAM capsule type (see {{iana-types}} for
+the value of the capsule type). This capsule has the same format and semantics
+as the DATAGRAM capsules, however intermediaries MUST NOT convert between
+RELIABLE_DATAGRAM caspules and other encodings of HTTP Datagrams, or vice versa.
+Endpoints MUST treat receipt of a RELIABLE_DATAGRAM capsule as equivalent to
+receipt of a DATAGRAM capsule with the same contents.
```

Comment:
I'm fairly sure that we didn't agree to this bit.

On the contrary, I think that several of us opposed it.  Can we cut it and take it up separately?

Reply:
I'm with Martin that I do not want to see this RELIABLE_DATAGRAM defined, especially not in this PR.

Especially given that capsules can't be modified by intermediaries, etc, I don't see the value in defining this here. If protocols want to define another capsule they use to transmit data over their stream of capsules, great—go ahead and define it there!

Reply:
I'd like us to discuss this in the context of this design team. The reason for that is that it has significant implications on future extensibility. I feel strongly that this spec needs enough of a toehold so that I can write a separate "datagram contexts" extension that has the "zero-latency extensibility" feature. The minimal toehold that I wrote initially was adding an UNEXTENDED_DATAGRAM capsule that has the same semantics as the DATAGRAM capsule but that intermediaries can't touch. Looking back I realized that this ended up being very similar to the RELIABLE_DATAGRAM that Ben asked for so I renamed it. I wrote up the way the extensibility toehold works in the "Creating a Demultiplexing Extension" appendix. I'm definitely open to other toehold that allow future extensibility, let me know if you have ideas.

Reply:
I see some generic utility of RELIABLE_DATAGRAM.  It's, well, reliable, and is not bound by MTU restrictions.  The only difference between RELIABLE_DATAGRAM and an unknown Capsule type is that an endpoint has to treat it the same as receiving a DATAGRAM Capsule, instead of dropping it.

I'm missing how having it enables zero-latency extensibility.

Reply:
I am also skeptical of this. 

I would ask if this question really needs to be resolved by this design team. Can't we discuss it in the WG?

Reply:
For me the extensibility properties are a fundamental part of the design, so the output of the design team needs to discuss this.

Reply:
> this is unordered but reliable

How is it unordered if it can't be converted to a QUIC DATAGRAM?

> That said, I agree with Martin that it doesn't belong in this PR. Let's work on one piece at a time

+1 

There's a lot here we have agreement on, so if RELIABLE_DATAGRAM is controversial let's break it out.  

I'm still missing how it enables or is required for 0-RTT extensions (which are a nice-to-have).  The appendix was vague.  @DavidSchinazi can you write down what you are thinking.

Reply:
The current draft (version -05) contains the datagram context extensibility system. I'm OK with us removing that from this draft (which is what this PR does), on the condition that I can still implement that as an extension to this draft. In particular, I need that extension to work without costing a round trip (what I've been calling "zero-latency extensibility"). Let me go into more detail as to why RELIABLE_DATAGRAM is important to that.

The datagram context extension works as follows: it is negotiated by adding a header to both request and response, and when both endpoints support it, it adds a context ID varint at the start of the HTTP Datagram Payload field.

Now, in order to achieve zero-latency extensibility, the client needs to be able to send datagrams before receiving the server headers in such a way that the server can correctly parse them (if you add the context ID and the server doesn't support the extension, you don't want the server to parse the context ID as part of the payload).

So here's my idea for how to accomplish that. The extension says the following:
- clients that support the extension are not allowed to send QUIC DATAGRAM frames nor DATAGRAM capsule before they've received the server headers
- the extension defines a new RELIABLE_DATAGRAM_EXT capsule that can be sent before and after receiving server headers
```
Extended Reliable Datagram Capsule {
  Type (i) = RELIABLE_DATAGRAM_EXT,
  Length (i),
  Context ID (i),
  HTTP Datagram Payload (..),
}
```
- the extension does not change the semantics of the RELIABLE_DATAGRAM capsule, that still does not contain a Context ID
```
Reliable Datagram Capsule {
  Type (i) = RELIABLE_DATAGRAM,
  Length (i),
  HTTP Datagram Payload (..),
}
```
- when both endpoints support the extension (after receiving headers) the extension changes the semantics of both the QUIC DATAGRAM frame and DATAGRAM capsule to include a Context ID at the start:
```
Datagram Capsule {
  Type (i) = DATAGRAM,
  Length (i),
  Context ID (i),
  HTTP Datagram Payload (..),
}
HTTP/3 Datagram { // contents of QUIC DATAGRAM frame
  Quarter Stream ID (i),
  Context ID (i),
  HTTP Datagram Payload (..),
}
```
- if the client wishes to use zero-latency extensibility, it can send RELIABLE_DATAGRAM and RELIABLE_DATAGRAM_EXT before receiving the server headers. Both of these will be parsed correctly by servers, whether or not the server supports the extension.

All of this relies on the fact that there is one bit that differentiates the DATAGRAM capsule from the RELIABLE_DATAGRAM capsule (the bit is in the capsule type), so that the extension can give the two capsules different semantics in a way that is backwards compatible.

What I like about this design is that the added complexity to the HTTP Datagram spec is as minimal as possible: it just requires two case statements instead of one in the switch statement that parses capsule types.

I'm OK removing the RELIABLE_DATAGRAM capsule from this PR, as long as we have some other mechanism that I can use to write my extension.

Reply:
I find this entire conversation a bit hard to follow, so forgive me if I've missed anything.

There's a claim that RELIABLE_DATAGRAM is controversial, and also a poor name. I haven't seen a good argument to remove it, merely the assertion that people do not want it. There are however a few use cases that have been described in its favor, so I personally would like to see it included (even if it has some other name), as per the discussion we've had at IETF and on list around the different MTU issues, as well as the H3->H2->H3 translation issues that it helps resolve.

A different issue that I'd like to raise (although I can live with this) is that I'm not really a fan of the extension-that-changes-the-semantics-of-DATAGRAM scheme that is being proposed here. I understand that is a natural consequence of the Datagram Capsule -> QUIC DATAGRAM encoding, but I'd personally really like to see the type varint make it into the QUIC DATAGRAM so that we can always unambiguously parse the datagrams, regardless of how we received them (capsule or otherwise). I acknowledge that this may not be the way others want this, and can live with the current (proposed) text.

Reply:
> There's a claim that RELIABLE_DATAGRAM is controversial, and also a poor name. I haven't seen a good argument to remove it, merely the assertion that people do not want it. 

This is the wrong standard. Things need consensus to be *included*, not removed.

In any case, a comment on this piece of the document is the wrong place to debate that, so the right thing to do is not have it in this PR and then discuss it separately.

Reply:
I disagree with your position that it is reasonable to omit things just because someone provides an unsubstantiated objection. Kindly please provide a technical argument for its removal, or we will not be able to build consensus in either direction.

Reply:
Maybe I don't understand what 0-RTT extensibility means.  Does it mean:

1. A client that supports an extension can try to use the extension before receiving confirmation that the server supports it
2. If the server supports the extension it will be processed correctly
3. If the server does not support the extension, the stream will not be torn down, but the extended datagrams could be dropped
---
If that's correct, can you build this without RELIABLE_DATAGRAM by:

1. Extension defines new Capsule type DATAGRAM_EXT
2. Client sends indication that it supports DATAGRAM_EXT in the CONNECT headers
3. Client sends DATAGRAM_EXT Capsules
3a.  Server supports extension, processes DATAGRAM_EXT Capsules
4a. Server confirms extension support
5a. Client receives acknowledgement that the extension is supported, switches to DATAGRAM and DATAGRAM Capsules
3b. Server silently drops DATAGRAM_EXT Capsules
4b. Server sends CONNECT response
5b. Client sees the server does not support the extension, uses standard DATAGRAM and DATAGRAM Capsules

I see that RELIABLE_DATAGRAM makes it possible for a server that supports the extension to assert than any DATAGRAM or DATAGRAM Capsule is only the extended flavor.  But how would an extended server handle a non-extended client in that case?  It seems like the extended server has to have a switch to know if the DATAGRAMs are extended or not, so I don't see any value in RELIABLE_DATAGRAM in this use case.  

Please help me understand if I'm missing something.

Reply:
The way this process is supposed to work is that *this* PR is supposed to include the points on which we had consensus, and that means when people object to things which clearly don't, they are removed and deferred for later discussion. If that's not going to happen, then we can just abandon this PR and go back to real time discussion.

Reply:
> A different issue that I'd like to raise (although I can live with this) is that I'm not really a fan of the extension-that-changes-the-semantics-of-DATAGRAM scheme that is being proposed here.

I am inclined to agree, but I see the solution as moving the demux higher in the protocol stack (eg: to CONNECT-UDP or CONNECT-IP) rather than down (into QUIC or H3 Datagram).  CONNECT-IP can say all DATAGRAMs on a CONNECT-IP stream have a specific format after the stream ID that includes "Context ID", "Format Type", etc, and this doesn't need to be negotiated at all.  I'd slightly prefer if those protocols used less generic concepts in doing so, but can live with it either way.

Reply:
ekr: You appear to have a particularly rigid definition of "the way this is supposed to work", and it does not match mine, and I rather suspect many other people's. I am not particularly interested in having a procedural discussion here, but rather working towards consensus on the document. To that end, I would like to see technical reasoning for everything that goes _in_ or gets taken _out_. I do not consider emotional appeals such as "I don't want this" to be valid.

Reply:
afrind: The one downside with your suggestion is it means the context sensitivity still exists, but is a property of the protocol rather than an extension. I think that makes logical sense, but doesn't fully solve the issue I was describing. It's probably the case that we don't necessarily have to solve it, as I mentioned before, I can live with the semantic change here.

Reply:
@afrind regarding your proposal in https://github.com/ietf-wg-masque/draft-ietf-masque-h3-datagram/pull/124#discussion_r753515209 the issue there is that it doesn't allow the client to send an unextended datagram in its first flight. For CONNECT-UDP that's really useful: the client wants to send its first UDP payload with the first flight and have the server still correctly parse that even if it doesn't support the extension.

@ekr and @achernya in terms of process here: it's clear that there isn't consensus to land this PR with RELIABLE_DATAGRAM, and there isn't consensus to land it without RELIABLE_DATAGRAM. We hadn't really discussed RELIABLE_DATAGRAM at our last in-person design team so I wasn't sure what people's thoughts on it were. Clearly it's something that warrants in-person discussion time. My plan is to keep the topic of RELIABLE_DATAGRAM as an open issue that we will resolve at our next in-person design team meeting. This PR won't get merged until this is resolved.

Reply:
> the issue there is that it doesn't allow the client to send an unextended datagram in its first flight.

I assume you mean 'unextended DATAGRAM Capsule'.  In that case, I think I understand, which is that a DATAGRAM Capsule can be converted to a QUIC DATAGRAM by an intermediary, and there's ambiguity at a server that supports the extension after it confirms support: when did the client receive confirmation and start sending extended QUIC DATAGRAMs?

The client can still send DATAGRAM Capsules in this case, and move to QUIC DATAGRAM after it knows whether the server supports the extension or not.  Is there a real use case for sending both extended and non-extended DATAGRAMs in 0-RTT, where we're willing to sacrifice HoL blocking resilience and possible flow control constraints for the extended ones, but not willing to do so for the non-extended ones?  Or maybe the use-cases are small but we think the cost of RELIABLE_DATAGRAM is low and maybe has other utility?

This brings me back here:

> I'm not really a fan of the extension-that-changes-the-semantics-of-DATAGRAM scheme that is being proposed here.

I think we've shown that it's pretty hard to negotiate changes to QUIC DATAGRAM formats on the fly.  Let's not do that.  The cost of easy extensibility is an additional byte (or two) of header information inside the datagram payload of protocols that need it.  If CONNECT-UDP needs to have a such a header I can live with that.

Reply:
QUIC DATAGRAM frames and DATAGRAM capsules need to have the exact same semantics, because intermediaries convert blindly between them so you can't know how they were originally sent. That's why my proposal forbids sending either when you support the extension and haven't yet received the peer's headers.

My personal opinion is that
- the cost of having RELIABLE_DATAGRAM for an implementation that doesn't care is very low
- RELIABLE_DATAGRAM has value in and of itself
- moving the complexity of contexts to their own extension allows defining different ways to do this down the road, whereas if we say that CONNECT-UDP starts with a varint we've committed to this context design forever

Let's discuss this in person on 11/29

Reply:
Congratulations, you have just collectively managed to do what we've learned through hard experience is a bad idea: a substantive technical discussion in a comment attached to a line of code on a pull request.

All I was asking for was for this to be taken to an issue.  Can we do that?

Reply:
Sure! I've filed #125 to discuss this.

Reply:
I've removed RELIABLE_DATAGRAM from this PR. We still have #111 to discuss RELIABLE_DATAGRAM and #125 to discuss zero-latency extensibility, so I'm now marking this comment thread as resolved.